src/solvers/refinement/string_constraint_generator_valueof.cpp

```diff
@@ -403,105 +403,124 @@ exprt string_constraint_generatort::add_axioms_for_correct_number_format(
   axioms.push_back(
     implies_exprt(correct, str.axiom_for_is_shorter_than(max_size)));
 
-  // forall 1 <= qvar < |str| . correct => is_digit_with_radix(str[qvar], radix)
-  symbol_exprt qvar=fresh_univ_index("number_format", index_type);
-  exprt is_digit=is_digit_with_radix(str[qvar], radix);
-  string_constraintt all_digits(
-    qvar, from_integer(1, index_type), str.length(), correct, is_digit);
-  axioms.push_back(all_digits);
+//  // forall 1 <= qvar < |str| . correct => is_digit_with_radix(str[qvar], radix)
+//  symbol_exprt qvar=fresh_univ_index("number_format", index_type);
+//  exprt is_digit=is_digit_with_radix(str[qvar], radix);
+//  string_constraintt all_digits(
+//    qvar, from_integer(1, index_type), str.length(), correct, is_digit);
+//  axioms.push_back(all_digits);
+
+  for(std::size_t index=1; index<max_size; ++index)
+  {
+    const exprt index_expr=from_integer(index, index_type);
+    /// index < length && correct => is_digit(str[index])
+    implies_exprt character_at_index_is_digit(
+      and_exprt(
+        binary_relation_exprt(index_expr, ID_lt, str.length()), correct),
+      is_digit_with_radix(str[index], radix));
+    axioms.push_back(character_at_index_is_digit);
+  }
```

Comment:
Yes. This will need refactoring when we try to tell whether the input string is formatted correctly.